extensions-core/druid-catalog/src/main/java/org/apache/druid/catalog/http/CatalogResource.java

```diff
@@ -0,0 +1,747 @@
+      return Actions.okWithVersion(newVersion);
+    }
+    catch (NotFoundException e) {
+      return Response.status(Response.Status.NOT_FOUND).build();
+    }
+    catch (OutOfDateException e) {
+      return Response
+          .status(Response.Status.BAD_REQUEST)
+          .entity(
+              Actions.error(
+                  Actions.DUPLICATE_ERROR,
+                  "The table entry not found or is older than the given version: reload and retry"))
+          .build();
+    }
+    catch (Exception e) {
+      return Actions.exception(e);
+    }
+  }
+
+  private Response addOrUpdateTableSpec(TableMetadata table)
+  {
+    try {
+      long newVersion = catalog.tables().create(table);
+      return Actions.okWithVersion(newVersion);
+    }
+    catch (DuplicateKeyException e) {
+      // Fall through
+    }
+    catch (Exception e) {
+      return Actions.exception(e);
+    }
+    try {
+      long newVersion = catalog.tables().update(table, 0);
+      return Actions.okWithVersion(newVersion);
+    }
+    catch (Exception e) {
+      return Actions.exception(e);
+    }
+  }
+
+  /**
+   * Update a table within the given schema.
+   *
+   * @param dbSchema The name of the Druid schema, which must be writable
+   *        and the user must have at least read access.
+   * @param name The name of the table definition to modify. The user must
+   *        have write access to the table.
+   * @param spec The new table definition.
+   * @param version An optional table version. If provided, the metadata DB
+   *        entry for the table must be at this exact version or the update
+   *        will fail. (Provides "optimistic locking.") If omitted (that is,
+   *        if zero), then no update conflict change is done.
+   * @param req the HTTP request used for authorization.
+   * @return the new version number of the table
+   */
+  @PUT
+  @Path("/tables/{dbSchema}/{name}")
+  @Consumes(MediaType.APPLICATION_JSON)
+  @Produces(MediaType.APPLICATION_JSON)
+  public Response updateTableDefn(
+      @PathParam("dbSchema") String dbSchema,
+      @PathParam("name") String name,
+      TableSpec spec,
+      @QueryParam("version") long version,
+      @Context final HttpServletRequest req
+  )
+  {
+
+    TableDefnRegistry tableRegistry = catalog.tableRegistry();
+    return incrementalUpdate(
+        TableId.of(dbSchema, name),
+        spec,
+        req,
+        (existing) -> tableRegistry.resolve(existing).merge(spec).spec()
+    );
+  }
+
+  private Response incrementalUpdate(
+      TableId tableId,
+      TableSpec newSpec,
+      @Context final HttpServletRequest req,
+      Function<TableSpec, TableSpec> action
+  )
+  {
+    Response response = authorizeTable(tableId, newSpec, req);
+    if (response != null) {
+      return response;
+    }
+    try {
+      long newVersion = catalog.tables().updatePayload(tableId, action);
+      return Actions.okWithVersion(newVersion);
+    }
+    catch (NotFoundException e) {
```

Comment:
Does it make sense to also throw and catch `OutOfDateException` as is done when force creating a table, so that the user can distinguish between the table not existing and them specifying the incorrect version?